use-cases/using-Intel-QAT-and-kata.md

```diff
@@ -0,0 +1,358 @@
+
+##Helpful Links before starting
+[https://github.com/intel/QAT_Engine](https://github.com/intel/QAT_Engine)
+
+[https://01.org/intel-quickassist-technology](https://01.org/intel-quickassist-technology)
+
+[https://github.com/intel/intel-device-plugins-for-kubernetes/](https://github.com/intel/intel-device-plugins-for-kubernetes/)
+
+[https://dpdk-docs.readthedocs.io/en/latest/cryptodevs/qat.html](https://dpdk-docs.readthedocs.io/en/latest/cryptodevs/qat.html)
+
+
+##Steps to enable Intel QAT in Kata Containers
+Some of the steps only have to be done once. Some have to be done every reboot. Other steps are required only when the host kernel changes.
+
+
+1. [Script Variables](#script-variables)
+    1. [Set environment variables (Every Reboot)](#script-variables-1)
+1. [Prepare the Host (Clear Linux)](#prepare-host)
+    1. [Identify which PCI bus the Intel QAT card is on](#prepare-1)
+    1. [Replace Clear Linux kernel with LTS kernel](#prepare-2a)
+    1. [Install necessary bundles for Clear Linux and disable kernel updates](#prepare-2)
+    1. [Allow insecure modules to load to host](#prepare-3)
+    1. [Download Intel QAT drivers](#prepare-4)
+    1. [Compile Intel QAT modules against host kernel](#prepare-5)
+    1. [Copy Intel QAT configuration files and enable vf’s](#prepare-6)
+    1. [Expose and Bind Intel QAT virtual functions to vfio-pci (Every reboot)](#prepare-7)
+    1. [Check Intel QAT virtual functions are enabled](#prepare-8)
+1. [Prepare Kata Containers](#prepare-kata)
+    1. [Download Kata Kernel Source](#prepare-kata-1)
+    1. [Build Kata Kernel](#prepare-kata-1a)
+    1. [Copy Kata Kernel](#prepare-kata-2)
+    1. [Prepare Kata root filesystem](#prepare-kata-3)
+    1. [Compile Intel QAT drivers for Kata Kernel and add to Kata rootfs](#prepare-kata-4)
+    1. [Copy Kata rootfs](#prepare-kata-4a)
+    1. [Update Kata configuration.toml to point to custom kernel and rootfs](#prepare-kata-5)
+1. [Verify Intel QAT works in a Docker\* Kata container](#verify-qat)
+    1. [Build OpenSSL Intel QAT engine container](#verify-qat-1)
+    1. [Test Intel QAT in Docker](#verify-qat-2)
+    1. [Troubleshooting](#verify-qat-3)
+1. [Optional Scripts](#optional-scripts)
+    1. [Verify Intel QAT card counters are incremented](#optional-scripts-1)
+    1. [Unbind all vf’s from vfio driver, reset, and rebind to the host](#optional-scripts-2)
+
+##Script variables<a name="script-variables"></a>
```

Comment:
No HTML in docs please - there are only a couple of instances where it is required but they are not required for links like this. See the tool above which auto-generates the TOC for you (HTML anchors are not required for it to work ;)

Reply:
Thanks for the pointer. I tried that tool out and it generated something like this. I used an online TOC generator.  Also, the tool is called check-markdown but the readme says kata-check-markdown.
```
$ ~/go/src/github.com/kata-containers/tests/cmd/check-markdown/check-markdown toc using-Intel-QAT-and-kata.md
)(#introduction
)(#helpful-links-before-startingting
)(#steps-to-enable-intel-qat-in-kata-containersners
)(#script-variables-les
)(#set-environment-variables-every-rebootry Reboot)
)(#prepare-the-host-clear-linuxLinux)
)(#identify-which-pci-bus-the-intel-qat-card-is-onrd is on
)(#install-necessary-bundles-for-clear-linux-and-disable-kernel-updates updates
)(#replace-clear-linux-kernel-with-lts-kernelS kernel
)(#allow-insecure-modules-to-load-to-host to host
)(#download-intel-qat-drivers drivers
)(#compile-intel-qat-drivers-against-host-kernel- kernel
)(#copy-intel-qat-configuration-files-and-enable-vfsable vf’s
)(#expose-and-bind-intel-qat-virtual-functions-to-vfio-pci-every-rebootry reboot)
)(#loop-through-all-the-virtual-functions-and-bind-to-the-vfio-driver
)(#verify-vfs-createdd
)(#check-intel-qat-virtual-functions-are-enabled enabled
)(#prepare-kata-containersners
)(#download-kata-kernel-sourcel Source
)(#build-kata-kernela Kernel
)(#change-kernel-config-options-to-support-qat
)(#copy-kata-kernela Kernel
)(#prepare-kata-root-filesystemlesystem
)(#make-sure-that-the-agent-versino-matches-the-installed-kata-runtime-version
)(#compile-intel-qat-drivers-for-kata-containers-kernel-and-add-to-kata-containers-rootfss rootfs
)(#usdm_drv-also-needs-to-be-copied-into-rootfs-modules-path-and-depmod-run
)(#note-usdm_drv-will-be-loaded-by-configuring-kata-kernel-boot-parameter-in-kata-configg
)(#ignore-any-errors-on-modulesbuiltin-and-modulesorderer
)(#create-kata-rootfs
)(#copy-kata-rootfsa rootfs
)(#update-kata-configurationtoml-to-point-to-custom-kernel-and-rootfsnd rootfs
)(#make-sure-usdm_drv-is-loaded-in-the-image
)(#change-the-default-kernel-and-rootfs
)(#verify-intel-qat-works-in-a-docker-kata-containers-containeriner
)(#build-openssl-intel-qat-engine-containeriner
)(#test-intel-qat-in-dockern Docker
)(#troubleshootingshooting
)(#optional-scriptsipts
)(#verify-intel-qat-card-counters-are-incrementedremented
)(#unbind-all-vfs-from-vfio-driver-reset-and-rebind-to-the-hostto the host
)(#loop-through-all-the-virtual-functions-and-unbind-the-vfio-driver-optional-to-resetT)

```

Reply:
Hi Eric,

The tool is called `kata-check-markdown` but does live in a `check-markdown` directory in the tests repo. I don't understand what you're seeing though:

```sh
$ cd $GOPATH/src/github.com/kata-containers/tests
$ make check-markdown
$ cd $GOPATH/src/github.com/kata-containers/documentation
$ ../tests/cmd/check-markdown/kata-check-markdown toc use-cases/using-Intel-QAT-and-kata.md
* [Table of Contents](#table-of-contents)
* [Introduction](#introduction)
    * [Helpful Links before starting](#helpful-links-before-starting)
    * [Steps to enable Intel QAT in Kata Containers](#steps-to-enable-intel-qat-in-kata-containers)
    * [Script variables](#script-variables)
        * [Set environment variables (Every Reboot)](#set-environment-variables-every-reboot)
    * [Prepare the Host (Clear Linux)](#prepare-the-host-clear-linux)
        * [Identify which PCI Bus the Intel QAT card is on](#identify-which-pci-bus-the-intel-qat-card-is-on)
        * [Install necessary bundles for Clear Linux and disable kernel updates](#install-necessary-bundles-for-clear-linux-and-disable-kernel-updates)
        * [Replace Clear Linux kernel with LTS kernel](#replace-clear-linux-kernel-with-lts-kernel)
        * [Allow insecure modules to load to host](#allow-insecure-modules-to-load-to-host)
        * [Download Intel QAT drivers](#download-intel-qat-drivers)
        * [Compile Intel QAT drivers against host kernel](#compile-intel-qat-drivers-against-host-kernel)
        * [Copy Intel QAT configuration files and enable VF’s](#copy-intel-qat-configuration-files-and-enable-vfs)
        * [Expose and Bind Intel QAT virtual functions to VFIO-PCI (Every reboot)](#expose-and-bind-intel-qat-virtual-functions-to-vfio-pci-every-reboot)
        * [Check Intel QAT virtual functions are enabled](#check-intel-qat-virtual-functions-are-enabled)
    * [Prepare Kata Containers](#prepare-kata-containers)
        * [Download Kata Kernel Source](#download-kata-kernel-source)
        * [Build Kata Kernel](#build-kata-kernel)
        * [Copy Kata Kernel](#copy-kata-kernel)
        * [Prepare Kata root filesystem](#prepare-kata-root-filesystem)
        * [Compile Intel QAT drivers for Kata Containers Kernel and add to Kata Containers rootfs](#compile-intel-qat-drivers-for-kata-containers-kernel-and-add-to-kata-containers-rootfs)
        * [Copy Kata rootfs](#copy-kata-rootfs)
        * [Update Kata configuration.toml to point to custom kernel and rootfs](#update-kata-configurationtoml-to-point-to-custom-kernel-and-rootfs)
    * [Verify Intel QAT works in a Docker Kata Containers container](#verify-intel-qat-works-in-a-docker-kata-containers-container)
    * [Build OpenSSL Intel QAT engine container](#build-openssl-intel-qat-engine-container)
        * [Test Intel QAT in Docker](#test-intel-qat-in-docker)
        * [Troubleshooting](#troubleshooting)
    * [Optional Scripts](#optional-scripts)
        * [Verify Intel QAT card counters are incremented](#verify-intel-qat-card-counters-are-incremented)
        * [Unbind all VF’s from VFIO driver, reset, and rebind to the host](#unbind-all-vfs-from-vfio-driver-reset-and-rebind-to-the-host)
```

Reply:
FTR, you don't *have* to use `kata-check-markdown` - use whatever tool you like, but just find one that doesn't add HTML to markdown docs :smile:

Reply:
I think I know what happened. I ran "go build" in the tests/cmd/check-markdown directory and it called it check-markdown. Also, as you figured out in IRC my doc was originally created on windows which proved challenging for the kata-check-markdown script.